### PR TITLE
Clean up the Proxies a bit, in prep for RealmInteger mods

### PR DESCRIPTION
**Pull request by bmeike**
I would like to be able to read the proxy code easily, so that I can make the (possibly somewhat complicated) changes for `RealmInteger`.  There should be absolutely no functional changes here: just syntax.

### CODE REVIEW

File: realm/realm-annotations-processor/src/main/java/io/realm/processor/RealmProxyClassGenerator.java [hunk 1]
```diff
@@ -65,43 +97,16 @@ public void generate() throws IOException, UnsupportedOperationException {
...
-        imports.add("org.json.JSONObject");
-        imports.add("org.json.JSONException");
-        imports.add("org.json.JSONArray");
-
         writer.emitImports(imports)
                 .emitEmptyLine();
 
         // Begin the class definition
-        writer.beginType(
+        writer.emitAnnotation("SuppressWarnings(\"all\")")
```
Comment (zaki50) [inline realm/realm-annotations-processor/src/main/java/io/realm/processor/RealmProxyClassGenerator.java]: This change should be discussed in another PR.

File: realm/realm-annotations-processor/src/main/java/io/realm/processor/RealmProxyClassGenerator.java [hunk 2]
```diff
@@ -599,7 +604,10 @@ private void emitCreateRealmObjectSchemaMethod(JavaWriter writer) throws IOExcep
                 EnumSet.of(Modifier.PUBLIC, Modifier.STATIC), // Modifiers
                 "RealmSchema", "realmSchema"); // Argument type & argument name
 
-        writer.beginControlFlow("if (!realmSchema.contains(\"" + this.simpleClassName + "\"))");
+        writer.beginControlFlow("if (realmSchema.contains(\"%s\"))", this.simpleClassName)
```
Comment (zaki50) [inline realm/realm-annotations-processor/src/main/java/io/realm/processor/RealmProxyClassGenerator.java]: I like this change! :+1:

File: realm/realm-annotations-processor/src/test/resources/io/realm/AllTypesRealmProxy.java [hunk 3]
```diff
@@ -399,26 +398,26 @@ protected final void copy(ColumnInfo rawSrc, ColumnInfo rawDst) {
     }
 
     public static RealmObjectSchema createRealmObjectSchema(RealmSchema realmSchema) {
+        if (realmSchema.contains("AllTypes")) {
```
Comment (cmelchior) [inline realm/realm-annotations-processor/src/test/resources/io/realm/AllTypesRealmProxy.java]: Any reason to add this? `createRealmObjectSchema` is only called if the Realm file is unversioned (=just created). So this will always be false? https://github.com/realm/realm-java/blob/master/realm/realm-library/src/main/java/io/realm/Realm.java#L461

File: realm/realm-annotations-processor/src/test/resources/io/realm/AllTypesRealmProxy.java [hunk 4]
```diff
@@ -399,26 +398,26 @@ protected final void copy(ColumnInfo rawSrc, ColumnInfo rawDst) {
...
-                AllTypesRealmProxy.createRealmObjectSchema(realmSchema);
-            }
-            realmObjectSchema.add("columnRealmList", RealmFieldType.LIST, realmSchema.get("AllTypes"));
-            return realmObjectSchema;
-        }
-        return realmSchema.get("AllTypes");
+            AllTypesRealmProxy.createRealmObjectSchema(realmSchema);
+        }
+        realmObjectSchema.add("columnObject", RealmFieldType.OBJECT, realmSchema.get("AllTypes"));
+        if (!realmSchema.contains("AllTypes")) {
```
Comment (cmelchior) [inline realm/realm-annotations-processor/src/test/resources/io/realm/AllTypesRealmProxy.java]: If you keep the if above, couldn't this just be removed then?

File: realm/realm-annotations-processor/src/test/resources/io/realm/AllTypesRealmProxy.java (hunk 3 above)
Comment (kneth) [inline realm/realm-annotations-processor/src/test/resources/io/realm/AllTypesRealmProxy.java]: Yes, if the cyclic model classes, and `createRealmObjectSchema()` calls itself. @beeender has a PR which changes that.

File: realm/realm-annotations-processor/src/test/resources/io/realm/AllTypesRealmProxy.java (hunk 4 above)
Comment (kneth) [inline realm/realm-annotations-processor/src/test/resources/io/realm/AllTypesRealmProxy.java]: You are tricked into thinking this since `AllTypes` has a field of type `AllTypes`. Of course, we can detect this in the proxy generator, but I can not sure it is worth doing. `createRealmObjectSchema()` is only called when initializing. Optimizing it and have a more complex proxy generator seems to be the wrong way to go.

File: realm/realm-annotations-processor/src/main/java/io/realm/processor/RealmProxyClassGenerator.java [hunk 5]
```diff
@@ -1225,10 +1230,12 @@ private void emitInsertListMethod(JavaWriter writer) throws IOException {
 
         writer.beginControlFlow("while (objects.hasNext())")
                 .emitStatement("object = (%s) objects.next()", qualifiedClassName);
-        writer.beginControlFlow("if(!cache.containsKey(object))");
+        writer.beginControlFlow("if (cache.containsKey(object))")
```
Comment (bmeike) [inline realm/realm-annotations-processor/src/main/java/io/realm/processor/RealmProxyClassGenerator.java]: I really don't know.  My guess would be that this makes not perceptible difference in execution time.  It almost certainly shaves a few bytes off the size of the executable, though.

I agree that it would be nice to have metrics for time to construct the proxy and time to run common proxy operations.

**Comment by bmeike** [comment]
Moved the SupressWarnings to a separate PR.  Can I get an "amen"?

File: realm/realm-annotations-processor/src/main/java/io/realm/processor/RealmProxyClassGenerator.java (hunk 5 above)
Comment (kneth) [inline realm/realm-annotations-processor/src/main/java/io/realm/processor/RealmProxyClassGenerator.java]: You could compare results from our micro benchmark.

Comment (cmelchior) [inline realm/realm-annotations-processor/src/main/java/io/realm/processor/RealmProxyClassGenerator.java]: I would be very surprised if this change was detectable even by the micro benchmarks. Anything below ~20% change is hidden by variance
